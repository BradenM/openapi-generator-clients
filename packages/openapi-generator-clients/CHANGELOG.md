# Changelog

## [0.1.1](https://github.com/BradenM/openapi-generator-clients/compare/openapi-generator-clients-v0.1.0...openapi-generator-clients-v0.1.1) (2023-02-13)


### Features

* **core:** Add cli entrypoint. ([cde5e51](https://github.com/BradenM/openapi-generator-clients/commit/cde5e51c0566d9b201de8a6be37e2299a103b2b1))
* **core:** Index exports. ([79cb4b8](https://github.com/BradenM/openapi-generator-clients/commit/79cb4b869fc22889aab2dadfc5d07785609ec61c))
* **core:** Refactor and reimplement generate as core build ([25b0851](https://github.com/BradenM/openapi-generator-clients/commit/25b085160348fb6f6f581b3fd0cbbe3900fb715c))
* **core:** Scaffold openapi-generator-clients core package. ([6268f33](https://github.com/BradenM/openapi-generator-clients/commit/6268f3317d001b5002f985cbff4c39cac2f45555))
* **core:** Support template level root dir. ([9cab69a](https://github.com/BradenM/openapi-generator-clients/commit/9cab69ad6ae9850e2875d73033f65242ad9af383))
* **core:** Temp workspace provider implementation. ([be6c90c](https://github.com/BradenM/openapi-generator-clients/commit/be6c90cc65fef3782c64b40ce8ba2cd4b1852331))


### Bug Fixes

* **core:** Downgrade rambdax to 8.1.0 ([72c90e2](https://github.com/BradenM/openapi-generator-clients/commit/72c90e2a58b1c60e4b721bdfae70f4b3eef807b4))
* **core:** Remove top level await in cli for cjs support ([5c5e866](https://github.com/BradenM/openapi-generator-clients/commit/5c5e866912d7909d630ffae0993484c0aff6aa59))
* **core:** Template type errors. ([ed9fd1e](https://github.com/BradenM/openapi-generator-clients/commit/ed9fd1e5153f23d709edd2a07cbcfb71dfe32050))
* **core:** Utilize cross cjs/esm resolve utils in build. ([9d729c7](https://github.com/BradenM/openapi-generator-clients/commit/9d729c78578f61a3d31c377eedfcc51d298992db))
* **pkg:** Add directory to core package. ([4ede56e](https://github.com/BradenM/openapi-generator-clients/commit/4ede56e35e202f417d272605108b024d8d49cf27))
