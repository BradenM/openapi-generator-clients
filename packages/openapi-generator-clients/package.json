{
	"name": "openapi-generator-clients",
	"type": "module",
	"version": "0.1.1",
	"description": "OpenAPI Generator CLI wrapper for managing multiple versioned client packages, versioned generators, and partial template overrides.",
	"author": "Braden Mars <bradenmars@bradenmars.me>",
	"license": "MIT",
	"homepage": "https://github.com/BradenM/openapi-generator-clients#readme",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/BradenM/openapi-generator-clients.git",
		"directory": "packages/openapi-generator-clients"
	},
	"bugs": "https://github.com/BradenM/openapi-generator-clients/issues",
	"keywords": [],
	"sideEffects": false,
	"main": "./dist/index.cjs",
	"module": "./dist/index.mjs",
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"require": "./dist/index.cjs",
			"import": "./dist/index.mjs"
		}
	},
	"types": "./dist/index.d.ts",
	"typesVersions": {
		"*": {
			"*": [
				"./dist/*",
				"./dist/index.d.ts"
			]
		}
	},
	"files": [
		"dist"
	],
	"bin": {
		"openapi-generator-clients": "./dist/cli.mjs"
	},
	"engines": {
		"node": ">=v16.20.2"
	},
	"scripts": {
		"build": "unbuild",
		"dev": "unbuild --stub",
		"prepublishOnly": "nr build",
		"start": "tsx src/index.ts",
		"test": "vitest",
		"typecheck": "tsc --noEmit"
	},
	"dependencies": {
		"@antfu/utils": "^0.7.8",
		"@openapi-generator-clients/config": "workspace:*",
		"@openapi-generator-clients/types": "workspace:*",
		"@openapi-generator-clients/utils": "workspace:*",
		"@openapitools/openapi-generator-cli": "^2.14.0",
		"consola": "^2.15.3",
		"execa": "^7.2.0",
		"fast-glob": "^3.3.2",
		"fs-extra": "^11.2.0",
		"immutable": "^4.3.7",
		"minimist": "^1.2.8",
		"pathe": "^1.1.2",
		"rambdax": "8.1.0",
		"type-fest": "^3.13.1"
	}
}
