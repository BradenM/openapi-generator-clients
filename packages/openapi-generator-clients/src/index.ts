export * from './build'
export * from '@openapi-generator-clients/types'
export {defineClients, defineGenerator} from '@openapi-generator-clients/config'
