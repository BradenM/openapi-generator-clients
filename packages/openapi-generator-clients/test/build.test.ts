import {vi, it, describe, expect, type SpyInstance, beforeEach} from 'vitest'
import type {
	Workspace,
	WorkspaceProvider,
} from '@openapi-generator-clients/types'
import {
	ClientRecord,
	GeneratorRecord,
	TemplateConfigRecord,
} from '@openapi-generator-clients/config'
import {buildClient} from 'openapi-generator-clients'

interface Context {
	workspaceProvider: WorkspaceProvider
	providerSpy: {
		create: SpyInstance
		addTemplatePath: SpyInstance
		setConfig: SpyInstance
	}
	inClient: ClientRecord
	workClient: ClientRecord
}

describe('buildClient', () => {
	const testWorkspace: Workspace = {
		rootPath: 'some-root',
		outputPath: 'test-output',
		configPath: 'test.config.json',
		templatesPath: 'test-templates',
	}

	beforeEach<Context>(async (ctx) => {
		ctx.workspaceProvider = {
			workspace: {...testWorkspace},
			async create() {
				return this
			},
			async addTemplatePath() {
				return this
			},
			async setConfig() {
				return this
			},
		}
		ctx.providerSpy = {
			create: vi.spyOn(ctx.workspaceProvider, 'create'),
			addTemplatePath: vi.spyOn(ctx.workspaceProvider, 'addTemplatePath'),
			setConfig: vi.spyOn(ctx.workspaceProvider, 'setConfig'),
		}
		ctx.inClient = new ClientRecord({
			name: 'test',
			generator: GeneratorRecord().mergeDeep({
				generatorVersion: 'v0',
				generatorName: 'test-generator',
				output: 'test-output',
			}),
			templateVersion: {
				version: 'v0',
				path: 'test-generator',
				config: TemplateConfigRecord().mergeDeep({generatorVersion: 'v0'}),
			},
			templateConfig: TemplateConfigRecord().merge({generatorVersion: 'v0'}),
			template: {name: 'test-template', root: 'path'},
		})
		ctx.workClient = await buildClient(ctx.inClient, {
			clients: [],
			overwrite: false,
			rootDir: '.',
			additionalArgs: [],
			workspaceProvider: ctx.workspaceProvider,
		})
	})

	it<Context>('creates expected workspace.', async (ctx) => {
		expect(ctx.workClient.toJS()).toMatchObject({
			workspace: testWorkspace,
		})
	})

	it<Context>('creates expected work client.', async (ctx) => {
		expect(ctx.workClient.toJS()).toMatchSnapshot('work client')
	})

	it<Context>('creates expected oas config.', async (ctx) => {
		expect(ctx.workClient.oasConfig.toJS()).toMatchSnapshot('oas config')
	})

	it<Context>('creates workspace and adds template paths.', async (ctx) => {
		expect(ctx.providerSpy.create).toHaveBeenCalledTimes(1)
		expect(ctx.providerSpy.addTemplatePath).toHaveBeenCalledTimes(2)
		expect(ctx.providerSpy.create.mock.lastCall).toStrictEqual([
			'test-output',
			false,
		])
		expect(ctx.providerSpy.addTemplatePath.mock.lastCall).toEqual([
			'test-generator',
		])
	})

	it<Context>('writes expected oas config to workspace config.', async (ctx) => {
		expect(ctx.providerSpy.setConfig).toHaveBeenCalledTimes(1)
		expect(ctx.providerSpy.setConfig.mock.calls).toMatchSnapshot(
			'workspace oas config.',
		)
	})
})
