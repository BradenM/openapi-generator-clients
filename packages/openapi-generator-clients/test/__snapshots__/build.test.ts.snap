// Vitest Snapshot v1

exports[`buildClient > creates expected oas config. > oas config 1`] = `
{
  "$schema": "node_modules/@openapitools/openapi-generator-cli/config.schema.json",
  "generator-cli": {
    "generators": {
      "test": {
        "additionalProperties": {
          "disallowAdditionalPropertiesIfNotPresent": undefined,
          "legacyDiscriminatorBehavior": false,
          "npmName": undefined,
          "platform": undefined,
          "supportsES6": true,
          "useInversify": undefined,
          "useObjectParameters": undefined,
          "usePromise": true,
          "useRxJS6": true,
          "withInterfaces": true,
        },
        "generatorName": "test-generator",
        "gitHost": "github.com",
        "gitRepoId": "",
        "gitUserId": "",
        "glob": undefined,
        "inputSpec": undefined,
        "legacyDiscriminatorBehavior": false,
        "output": "test-output",
        "removeOperationIdPrefix": true,
        "templateDir": "test-templates",
      },
    },
    "version": "6.0.0-beta",
  },
  "spaces": 2,
}
`;

exports[`buildClient > creates expected work client. > work client 1`] = `
{
  "generator": {
    "additionalArgs": [],
    "additionalProperties": {
      "disallowAdditionalPropertiesIfNotPresent": undefined,
      "legacyDiscriminatorBehavior": false,
      "npmName": undefined,
      "platform": undefined,
      "supportsES6": true,
      "useInversify": undefined,
      "useObjectParameters": undefined,
      "usePromise": true,
      "useRxJS6": true,
      "withInterfaces": true,
    },
    "drop": [],
    "generatorName": "test-generator",
    "generatorVersion": "v0",
    "gitHost": "github.com",
    "gitRepoId": "",
    "gitUserId": "",
    "glob": undefined,
    "inputSpec": undefined,
    "legacyDiscriminatorBehavior": false,
    "output": "test-output",
    "removeOperationIdPrefix": true,
    "templateDir": "test-templates",
  },
  "name": "test",
  "template": {
    "name": "test-template",
    "root": "path",
  },
  "templateConfig": {
    "additionalArgs": [],
    "drop": [],
    "generatorVersion": "v0",
  },
  "templateVersion": {
    "config": {
      "additionalArgs": [],
      "drop": [],
      "generatorVersion": "v0",
    },
    "path": "test-generator",
    "version": "v0",
  },
  "workspace": {
    "configPath": "test.config.json",
    "outputPath": "test-output",
    "rootPath": "some-root",
    "templatesPath": "test-templates",
  },
}
`;

exports[`buildClient > writes expected oas config to workspace config. > workspace oas config. 1`] = `
[
  [
    {
      "additionalProperties": {
        "disallowAdditionalPropertiesIfNotPresent": undefined,
        "legacyDiscriminatorBehavior": false,
        "npmName": undefined,
        "platform": undefined,
        "supportsES6": true,
        "useInversify": undefined,
        "useObjectParameters": undefined,
        "usePromise": true,
        "useRxJS6": true,
        "withInterfaces": true,
      },
      "generatorName": "test-generator",
      "gitHost": "github.com",
      "gitRepoId": "",
      "gitUserId": "",
      "glob": undefined,
      "inputSpec": undefined,
      "legacyDiscriminatorBehavior": false,
      "output": "test-output",
      "removeOperationIdPrefix": true,
      "templateDir": "test-templates",
    },
  ],
]
`;
